### PR TITLE
ports GAGS

### PR DESCRIPTION
**Pull request by tattax**
# Document the changes in your pull request

What is GAGS?

GAGS are a system to make auto gen'd sprites that use a greyscale to change colors instead of having the same sprite with different colors copypasted all over again, which makes sprites easier to make and update, see more here: https://github.com/tgstation/tgstation/pull/58112

I've ported every javascript reader to make it easier to port / implement future GAGS

# Spriting

<details>
  <summary>atmos canisters</summary>

roundstart atmos setup

image

toxins storage

image

  
</details>

<details>
  <summary>PDAs</summary>
  
inspired from: https://github.com/tgstation/tgstation/pull/66369

readds role PDAs, does not change pda painter just yet. from left to right, starting top left
sig tech - chap - engi - shaft miner - scientist
mime - clown - robo - chemist - MD
using different programs to show off new sprites for the overlays, the pinpointer overlay should be VISIBLY better
image
  
</details>

<details>
  <summary>carps</summary>
  
magicarps
aegm0setiX

chaoscarps
E0EiqOkStf

</details>

<details>
  <summary>PDA painter</summary>

from: https://github.com/tgstation/tgstation/pull/66076

image

</details>

<details>
  <summary>clothes</summary>
  
jumpsuits
from: https://github.com/tgstation/tgstation/pull/65264

image

shoes based on : https://github.com/tgstation/tgstation/pull/69134

image

sombreros (not default colors to show admintool)
image

</details>

# Wiki Documentation

none, just change the current icons that are in the spriting section to new ones

:cl:   
rscadd: admins can change colors on a specific item if it uses the GAGS system
rscadd: you can now use right click -> toggle flashlight to toggle PDA flashlight
imageadd: changed some sprites to use GAGS
/:cl:

### CODE REVIEW

**Comment by neeshacark** [comment]
this is NOT the pr i hoped it was

**Comment by azzzertyy** [comment]
thank you

**Comment by azzzertyy** [comment]
> this is NOT the pr i hoped it was

no this is goated you just dont know it

**Comment by neeshacark** [comment]
the changed wirecutters sprite might look a bit odd when all the other tool sprites are the same, is there some for those too? with or without the color?

**Comment by tattax** [comment]
> the changed wirecutters sprite might look a bit odd when all the other tool sprites are the same, is there some for those too? with or without the color?

yeah but it's easier for me to port it to switch it to use the GAGS format and i think it looks nicer, other tools don't use GAGS and the screwdriver wasn't updated beforehand but i could do that if you want

**Review by JohnFulpWillard — APPROVED** [review]

**Review by ToasterBiome — CHANGES_REQUESTED** [review]
- these canisters dont look very good, i assume these are the ones that are supposed to be in box. for reference:
image

These should all look at least reasonably close to how they look before, only using the new system.

- The PDAs are just bad. look like e-readers. Not halting this over that because me and cuackles have been cooking new PDAs.

- wirecutters should use old sprites and colors but with the new system

**Comment by tattax** [comment]
> * wirecutters should use old sprites and colors but with the new system

the old sprites are bad, you couldn't tell those were wirecutters and everything was all over the place + it was off centered

**Comment by tattax** [comment]
also 
image

**Comment by monster860** [comment]
Dear god, this is going to be FUUUUUUUN to make work with the replay system

**Comment by azzzertyy** [comment]
we have phones, keep the pdas the same

**Comment by cuackles** [comment]
honestly looking at all those sprites most of them seem like a direct downgrade, very few actually seem that would improve, and the amount of sprites u r porting is waaaay to much for 1 single pr, i would recommend dividing the pr into different part, first the system them the sprites that would use the system , also if someone of the art team makes a point in a change request it isnt arguable, its part of the change request and should be changed as per request of the reviewer

**Comment by iamthesilly** [comment]
Wrenching a canister makes its pressure indicator disappear and the overlay for it being wrenched down doesn't show up
image

Also PDAs are broken
image

Toolboxes seem to be discolored as well
image

**Review by cuackles — CHANGES_REQUESTED** [review]
after some thinking, i want to ask for this pr to only touch jumpsuit sprites, this system is a big change spritewise and i wanna see it working properly before merging 3 iconbot pages worth of sprites, so i need that the only sprites this pr brings or changes be jumpsuits, that way we can see how well the system works without having to focus if it works on cans but not on pdas, or if it breaks any sprite, if it works properly and it doesnt make jumpsuits look bad, i will give my thumbs up to keep integrating more things with this system

**Comment by ShoziOffical** [comment]
god bless this will make it easier to port other GAGS stuffs

**Comment by JohnFulpWillard** [comment]
I think it is worth the sprite change now that we are porting GAGS as implementing as much as we can of it would be useful for us in the long run, both in ensuring we won't forget to do it, and so contributors know of this feature and have many examples to use when trying to add new ones. Sprites can be reworked/reverted later (using GAGS) by anyone willing perhaps.

File: code/modules/modular_computers/computers/item/pda/pda_presets.dm [hunk 1]
```diff
@@ -1,117 +1,211 @@
...
+/obj/item/modular_computer/tablet/pda/preset/security/detective
+	greyscale_colors = "#805A2F#990202"
+
+/obj/item/modular_computer/tablet/pda/preset/security/warden
+	greyscale_config = /datum/greyscale_config/tablet/stripe_split
+	greyscale_colors = "#EA3232#0000CC#363636"
+
+/obj/item/modular_computer/tablet/pda/preset/security/warden/Initialize(mapload)
+	starting_files |= list(
+		new /datum/computer_file/program/secureye,
```
Comment (ToasterBiome) [inline code/modules/modular_computers/computers/item/pda/pda_presets.dm:199]: Why are you giving them secureye here? Where was this before?

File: code/modules/modular_computers/computers/item/pda/pda_presets.dm [hunk 2]
```diff
@@ -1,117 +1,211 @@
...
+	greyscale_config = /datum/greyscale_config/tablet/stripe_thick
+	greyscale_colors = "#a80001#5C070F#000000"
 
-/obj/item/modular_computer/tablet/pda/preset/basic/syndicate/Initialize(mapload)
+/obj/item/modular_computer/tablet/pda/preset/syndicate/Initialize(mapload)
 	. = ..()
 	obj_flags |= EMAGGED //starts emagged
 	starting_files |= list(
-		new /datum/computer_file/program/bomberman
+		new /datum/computer_file/program/bomberman,
```
Comment (ToasterBiome) [inline code/modules/modular_computers/computers/item/pda/pda_presets.dm:212]: adding commas here is stupid

File: code/modules/reagents/reagent_containers/blood_pack.dm [hunk 3]
```diff
@@ -77,15 +77,13 @@
...
-	if(!labelled)
-		if(blood_type)
-			name = "blood pack - [blood_type]"
-		else
-			name = "blood pack"
+/obj/item/reagent_containers/blood/update_name()
+	. = ..()
+	if(labelled)
+		return
+	name = "blood pack[blood_type ? " - [blood_type]" : ""]"
```
Comment (ToasterBiome) [inline code/modules/reagents/reagent_containers/blood_pack.dm]: This is unrelated to GAGS

File: code/modules/tgui/states.dm [hunk 4]
```diff
@@ -116,7 +116,7 @@
  *
  * return UI_state The state of the UI.
  */
-/mob/living/proc/shared_living_ui_distance(atom/movable/src_object, viewcheck = TRUE)
+/mob/living/proc/shared_living_ui_distance(atom/movable/src_object, viewcheck = TRUE, allow_tk = TRUE)
```
Comment (ToasterBiome) [inline code/modules/tgui/states.dm:119]: unrelated to GAGS

**Review by ToasterBiome — CHANGES_REQUESTED** [review]
Why does this PR have 300 changed files and a bunch of stuff completely unrelated to GAGS?

File: code/__DEFINES/subsystems.dm [hunk 5]
```diff
@@ -131,15 +131,16 @@
...
 #define INIT_ORDER_VIS				80
 #define INIT_ORDER_MATERIALS		76
 #define INIT_ORDER_RESEARCH			75
 #define INIT_ORDER_STATION			74
 #define INIT_ORDER_QUIRKS 			73
 #define INIT_ORDER_EVENTS			70
 #define INIT_ORDER_JOBS				65
-#define INIT_ORDER_MAPPING			60
-#define INIT_ORDER_TICKER			50
+#define INIT_ORDER_TICKER			55
```
Comment (ToasterBiome) [inline code/__DEFINES/subsystems.dm]: Why are you switching these

File: code/modules/jobs/job_types/assistant.dm [hunk 6]
```diff
@@ -63,5 +63,5 @@ Assistant
 
 	// This outfit is used by the assets SS, which is ran before the atoms SS
 	if (SSatoms.initialized == INITIALIZATION_INSSATOMS)
-	//	H.w_uniform?.update_greyscale()
+		H.w_uniform?.update_greyscale()
```
Comment (ToasterBiome) [inline code/modules/jobs/job_types/assistant.dm:65]: Why is this uncommented out?

**Review by ToasterBiome — CHANGES_REQUESTED** [review]

File: code/modules/modular_computers/computers/item/pda/pda_presets.dm (hunk 1 above)
Comment (tattax) [reply code/modules/modular_computers/computers/item/pda/pda_presets.dm:199]: it was, original one has secureye, it is just repathed so it may seem weird but it was there before

File: code/modules/modular_computers/computers/item/pda/pda_presets.dm (hunk 2 above)
Comment (tattax) [reply code/modules/modular_computers/computers/item/pda/pda_presets.dm:212]: it is good practice to avoid any errors in the future

File: code/modules/reagents/reagent_containers/blood_pack.dm (hunk 3 above)
Comment (tattax) [reply code/modules/reagents/reagent_containers/blood_pack.dm]: it is just a code improv that i did so i could rerun tests

File: code/modules/tgui/states.dm (hunk 4 above)
Comment (tattax) [reply code/modules/tgui/states.dm:119]: i needed to port some states and this is basically just a bug fix

Comment (tattax) [reply code/modules/tgui/states.dm:119]: (for gags tow ork)

File: code/modules/jobs/job_types/assistant.dm (hunk 6 above)
Comment (tattax) [reply code/modules/jobs/job_types/assistant.dm:65]: i commented it out when i ported it because we didn't have gags but now we do

File: code/__DEFINES/subsystems.dm (hunk 5 above)
Comment (tattax) [reply code/__DEFINES/subsystems.dm]: i changed it when i was testing to see why it didn't work the first time around, everything seems fine with these changes

File: code/__DEFINES/colors.dm [hunk 7]
```diff
@@ -247,7 +247,24 @@
 #define CIRCUIT_COLOR_MEDICAL "#00CCFF"
 #define CIRCUIT_COLOR_ENGINEERING "#F8D700"
 #define CIRCUIT_COLOR_SUPPLY "#C47749"
-
+#define CABLE_COLOR_BLUE 	"blue"
```
Comment (ToasterBiome) [inline code/__DEFINES/colors.dm]: Why is this file changed at all? It's just moving stuff.

File: code/__DEFINES/flags.dm [hunk 8]
```diff
@@ -44,12 +44,13 @@ GLOBAL_LIST_INIT(bitflags, list(1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 204
 /// should not get harmed if this gets caught by an explosion?
 #define PREVENT_CONTENTS_EXPLOSION_1 (1<<14)
 /// should the contents of this atom be acted upon
-#define RAD_PROTECT_CONTENTS_1 (1 << 15)
+#define RAD_PROTECT_CONTENTS_1 (1<<15)
```
Comment (ToasterBiome) [inline code/__DEFINES/flags.dm]: Why are these changed?

File: code/__DEFINES/subsystems.dm (hunk 5 above)
Comment (ToasterBiome) [inline code/__DEFINES/subsystems.dm]: Revert this change

**Review by ToasterBiome — CHANGES_REQUESTED** [review]
Monster still has to look at this

File: code/__DEFINES/colors.dm (hunk 7 above)
Comment (tattax) [reply code/__DEFINES/colors.dm]: it's just moving color defines next to each other there is no issue here

File: code/__DEFINES/flags.dm (hunk 8 above)
Comment (tattax) [reply code/__DEFINES/flags.dm]: consistency

Comment (tattax) [reply code/__DEFINES/flags.dm]: i was merging master and i changed it

**Review by ToasterBiome — CHANGES_REQUESTED** [review]
Art review:

- Canisters need to be closer to their original colors and designs
- Carps need to be brighter, closer to their original colors
- Shoes are ugly, I'd prefer to GAGSify our current shoe sprite.

**Review by ToasterBiome — DISMISSED** [review]
Please show the requested changes in action

**Comment by JohnFulpWillard** [comment]
@cuackles Do you still want this PR to only touch jumpsuits or did you talk over it or what?

**Comment by cuackles** [comment]
> @cuackles Do you still want this PR to only touch jumpsuits or did you talk over it or what?

kinda lost track of this due to time, since a lot of sprites have been changes i will like to take a careful look at it to make sure no sprites are accidentally deleted like it has happened on the past, about the sprites we can change without much of an issue, the pdas that biome made look very good, we can do the sombrero since its only like 1 sprite, i would like it so mqiib took a look at the gags affected clothing to make sure nothing dissapeared from his big change pr, and we can add the gaged canisters tho i would like to keep the old ones on files, at least until gags becomes a fully integrated system on yogs, i would like the shoes not to be affected just yet as the ones on this pr look very different from the ones we have
so tldr on what we can gags

- Pdas
- jumpsuits after having mqiib look at them to make sure nothing is lost
- cannisters but leaving the originals on the sprite files for the time being

**Comment by Mqiib** [comment]
Considering I can see the most recent sprites, the turtlenecks, in there, it seems the icons are all up to date with our current stuff. Consider this my pseudo-sprite-maintainer approval.

**Comment by JohnFulpWillard** [comment]
> until gags becomes a fully integrated system on yogs

This IS being added as a fully integrated system, and if you want to bring old sprites back then you can get them from an older branch or anything, I don't think we need to keep the sprites around esp since we might end up forgetting about it and they take up rsc size for nothing.

**Comment by tattax** [comment]
@cuackles @ToasterBiome

**Comment by Mqiib** [comment]
https://github.com/yogstation13/Yogstation/pull/21394
Me make conflict oopsies but it's just the female_top and female_full clothing masks at the bottom.

**Review by cuackles — APPROVED** [review]

File: code/__DEFINES/colors.dm [hunk 9]
```diff
@@ -269,7 +269,24 @@
...
+#define CABLE_COLOR_ORANGE 	"orange"
+	#define CABLE_HEX_COLOR_ORANGE 	COLOR_MOSTLY_PURE_ORANGE
+#define CABLE_COLOR_PINK 	"pink"
+	#define CABLE_HEX_COLOR_PINK 	COLOR_LIGHT_PINK
+#define CABLE_COLOR_RED 	"red"
+	#define CABLE_HEX_COLOR_RED 	COLOR_RED
+#define CABLE_COLOR_WHITE 	"white"
+	#define CABLE_HEX_COLOR_WHITE 	COLOR_WHITE
+#define CABLE_COLOR_YELLOW 	"yellow"
+	#define CABLE_HEX_COLOR_YELLOW 	COLOR_YELLOW
```
Comment (JohnFulpWillard) [inline code/__DEFINES/colors.dm]:
```suggestion
#define CABLE_COLOR_BLUE "blue"
	#define CABLE_HEX_COLOR_BLUE COLOR_STRONG_BLUE
#define CABLE_COLOR_BROWN "brown"
	#define CABLE_HEX_COLOR_BROWN COLOR_ORANGE_BROWN
#define CABLE_COLOR_CYAN "cyan"
	#define CABLE_HEX_COLOR_CYAN COLOR_CYAN
#define CABLE_COLOR_GREEN "green"
	#define CABLE_HEX_COLOR_GREEN COLOR_DARK_LIME
#define CABLE_COLOR_ORANGE "orange"
	#define CABLE_HEX_COLOR_ORANGE COLOR_MOSTLY_PURE_ORANGE
#define CABLE_COLOR_PINK "pink"
	#define CABLE_HEX_COLOR_PINK COLOR_LIGHT_PINK
#define CABLE_COLOR_RED "red"
	#define CABLE_HEX_COLOR_RED COLOR_RED
#define CABLE_COLOR_WHITE "white"
	#define CABLE_HEX_COLOR_WHITE COLOR_WHITE
#define CABLE_COLOR_YELLOW "yellow"
	#define CABLE_HEX_COLOR_YELLOW COLOR_YELLOW
```

File: code/__DEFINES/flags.dm [hunk 10]
```diff
@@ -42,6 +42,8 @@ GLOBAL_LIST_INIT(bitflags, list(1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 204
 #define NODECONSTRUCT_1				(1<<7)
 /// Prevent clicking things below it on the same turf eg. doors/ fulltile windows
 #define PREVENT_CLICK_UNDER_1		(1<<9)
+/// Can players recolor this in-game via vendors (and maybe more if support is added)?
+#define IS_PLAYER_COLORABLE_1 		(1<<10)
```
Comment (JohnFulpWillard) [inline code/__DEFINES/flags.dm]:
```suggestion
#define IS_PLAYER_COLORABLE_1 (1<<10)
```

File: code/__DEFINES/obj_flags.dm [hunk 11]
```diff
@@ -51,21 +51,37 @@
...
-#define VOICEBOX_DISABLED 		(1<<9) // The voicebox is currently turned off.
-#define SHOWEROKAY				(1<<10)	//prevents you from being stupid if you shower in them
-#define SCAN_REAGENTS 			(1<<11) // Allows helmets and glasses to scan reagents.
-#define MASKEXTENDRANGE			(1<<12) //! For masks, allows you to breathe from internals on adjecent tiles
+/// SUIT and HEAD items which stop lava from hurting the wearer
+#define LAVAPROTECT (1<<0)
+/// SUIT and HEAD items which stop pressure damage.
+/// To stop you taking all pressure damage you must have both a suit and head item with these flags. First one is high pressure (fires), second one is low (space).
+#define STOPSHIGHPRESSURE		(1<<1) 
+#define STOPSLOWPRESSURE		(1<<2)
```
Comment (JohnFulpWillard) [inline code/__DEFINES/obj_flags.dm]:
```suggestion
#define STOPSHIGHPRESSURE (1<<1) 
#define STOPSLOWPRESSURE (1<<2)
```

File: code/__DEFINES/subsystems.dm [hunk 12]
```diff
@@ -137,6 +137,7 @@
 #define INIT_ORDER_INPUT			85
 #define INIT_ORDER_SOUNDS			83
 #define INIT_ORDER_INSTRUMENTS		82
+#define INIT_ORDER_GREYSCALE	 	81
```
Comment (JohnFulpWillard) [inline code/__DEFINES/subsystems.dm]:
```suggestion
#define INIT_ORDER_GREYSCALE 81
```

File: code/_compile_options.dm [hunk 13]
```diff
@@ -21,11 +21,14 @@
...
 
 // If this is uncommented, we do a single run though of the game setup and tear down process with unit tests in between
-// #define UNIT_TESTS
+//#define UNIT_TESTS
 
 // If defined, we will NOT defer asset generation till later in the game, and will instead do it all at once, during initiialize
 //#define DO_NOT_DEFER_ASSETS
 
+// Uncomment to run runtimestation (less time to compile)
+//#define LOWMEMORYMODE
```
Comment (JohnFulpWillard) [inline code/_compile_options.dm:30]: This is in map list and shouldn't be here, probably.

File: code/datums/elements/update_icon_updates_onmob.dm [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+	SIGNAL_HANDLER
+
+	if(ismob(target.loc))
+		var/mob/mob = target.loc
+		if(mob.is_holding(target))
+			mob.update_inv_hands()
+		else
+			mob.update_clothing((target.slot_flags|update_flags))
+			if(update_body)
+				mob.update_body()
```
Comment (JohnFulpWillard) [inline code/datums/elements/update_icon_updates_onmob.dm]:
```suggestion
	if(!ismob(target.loc))
		return
	var/mob/mob = target.loc
	if(mob.is_holding(target))
		mob.update_inv_hands()
	else
		mob.update_clothing((target.slot_flags|update_flags))
		if(update_body)
			mob.update_body()
```

File: code/datums/greyscale/_greyscale_config.dm [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+	return ..()
+
+/datum/greyscale_config/process(seconds_per_tick)
+	if(!Refresh(loadFromDisk=TRUE))
+		return
+	if(!live_edit_types)
+		return
+	for(var/atom/thing in world)
+		if(live_edit_types[thing.type])
+			thing.update_greyscale()
```
Comment (JohnFulpWillard) [inline code/datums/greyscale/_greyscale_config.dm:80]: im taking this upstream this seems crazy

File: code/datums/greyscale/config_types/greyscale_configs.dm [hunk 16]
```diff
@@ -0,0 +1,1042 @@
...
+/datum/greyscale_config/party_hat
+	name = "Party Hat"
+	icon_file = 'icons/obj/clothing/head/costume.dmi'
+	json_config = 'code/datums/greyscale/json_configs/party_hat.json'
+
+/datum/greyscale_config/party_hat_worn
+	name = "Party Hat Worn"
+	icon_file = 'icons/mob/clothing/head/costume.dmi'
+	json_config = 'code/datums/greyscale/json_configs/party_hat.json'
+*/
```
Comment (JohnFulpWillard) [inline code/datums/greyscale/config_types/greyscale_configs.dm:301]: can you please remove all the unused configs

File: code/datums/greyscale/json_configs/garden_gnome.json [hunk 17]
```diff
@@ -0,0 +1,126 @@
+{
```
Comment (JohnFulpWillard) [inline code/datums/greyscale/json_configs/garden_gnome.json]: Like this is a lot of json files for things that aren't items that exist in the codebase, a lot of dead code that might never be used.

File: code/datums/materials/_material.dm [hunk 18]
```diff
@@ -44,50 +79,97 @@ Simple datum which is instanced once per type and is used for every object of sa
...
 
 ///This proc is called when the material is removed from an object.
-/datum/material/proc/on_removed(atom/source, material_flags)
-	if(!(material_flags & MATERIAL_NO_COLOR)) //Prevent changing things with pre-set colors, to keep colored toolboxes their looks for example
+/datum/material/proc/on_removed(atom/source, amount, material_flags)
+	if(material_flags & MATERIAL_COLOR) //Prevent changing things with pre-set colors, to keep colored toolboxes their looks for example
 		if(color)
 			source.remove_atom_colour(FIXED_COLOUR_PRIORITY, color)
+		if(texture_layer_icon_state)
+//			source.remove_filter("material_texture_[name]")
```
Comment (JohnFulpWillard) [inline code/datums/materials/_material.dm]: Can this filter stuff be added now that we got the plane cube PR?

**Review by JohnFulpWillard — CHANGES_REQUESTED** [review]
well hey since we're waiting on maptainers, here's some reviews

**Review by neeshacark — APPROVED** [review]
john wants map review idk why i dont see the request

its fine

File: code/datums/greyscale/_greyscale_config.dm (hunk 15 above)
Comment (JohnFulpWillard) [reply code/datums/greyscale/_greyscale_config.dm:80]: Nevermind this is all just debug stuff and doesn't run in normal gameplay.

File: code/datums/materials/_material.dm (hunk 18 above)
Comment (tattax) [reply code/datums/materials/_material.dm]: prob

**Review by JohnFulpWillard — APPROVED** [review]
stop calling the mule asshole im using it

**Comment by MommaDragon** [comment]
Obelisk says this might've broken the Path Markers, which has in turn made Mining impossible due to lack of sight (the light is sprite-based for some reason)
